pkg/segment/writer/metrics/wal/wal.go

```diff
@@ -0,0 +1,303 @@
+		return nil, false, err
+	}
+
+	it.readBuf = toputils.ResizeSlice(it.readBuf, int(blockSize-4))
+	_, err = io.ReadFull(it.fd, it.readBuf)
+	if err != nil {
+		log.Errorf("WalIterator Next: failed to read block data of size %d from file %s: %v", blockSize, it.fd.Name(), err)
+		return nil, false, err
+	}
+
+	calculatedChecksum := crc32.ChecksumIEEE(it.readBuf)
+	if calculatedChecksum != checksum {
+		log.Errorf("WalIterator Next: checksum mismatch! Calculated: %v, Expected: %v", calculatedChecksum, checksum)
+		return nil, false, errors.New("checksum mismatch")
+	}
+
+	blockBuf := bytes.NewReader(it.readBuf)
+	err = it.decodeWALBlock(blockBuf)
+
+	if err != nil {
+		log.Errorf("WalIterator Next: dataDecompression failed: %v", err)
+		return nil, false, err
+	}
+
+	it.currentIndex++
+	return &it.readDps[it.currentIndex-1], true, nil
+}
+
+func (it *WalIterator) Close() error {
+	if it.fd != nil {
+		return it.fd.Close()
+	}
+	return errors.New("file descriptor is nil")
+}
+
+func (w *WAL) encodeWALBlock(dps []WalDatapoint) error {
+	N := uint32(len(dps))
+	if N == 0 {
+		log.Warn("EncodeWALBlock: received empty data points slice")
+		return errors.New("empty data points")
+	}
+
+	w.rawBlockBuf.Reset()
+	err := binary.Write(w.rawBlockBuf, binary.LittleEndian, N)
+	if err != nil {
+		log.Errorf("EncodeWALBlock: failed to write datapoint count: %v", err)
+		return err
+	}
+
+	for _, dp := range dps {
+		if err := binary.Write(w.rawBlockBuf, binary.LittleEndian, dp.timestamp); err != nil {
+			log.Errorf("EncodeWALBlock: failed to write timestamp : %v err : %v", dp.timestamp, err)
+			return err
+		}
+	}
+
+	for _, dp := range dps {
+		if err := binary.Write(w.rawBlockBuf, binary.LittleEndian, dp.dpVal); err != nil {
+			log.Errorf("EncodeWALBlock: failed to write dpVal : %v err : %v", dp.dpVal, err)
+			return err
+		}
+	}
+
+	for _, dp := range dps {
+		if err := binary.Write(w.rawBlockBuf, binary.LittleEndian, dp.tsid); err != nil {
+			log.Errorf("EncodeWALBlock: failed to write tsid : %v err : %v", dp.tsid, err)
+			return err
+		}
+	}
+
+	w.encodeBuf = encoder.EncodeAll(w.rawBlockBuf.Bytes(), w.encodeBuf[:0])
+	return nil
+}
+
+func (it *WalIterator) decodeWALBlock(blockBuf *bytes.Reader) error {
+	it.readBlockBuf = toputils.ResizeSlice(it.readBlockBuf, blockBuf.Len())
+	_, err := blockBuf.Read(it.readBlockBuf)
+
+	if err != nil {
+		log.Errorf("decodeWALBlock: failed to read from blockBuf into blockBuf: %v", err)
+		return err
+	}
+
+	it.readBuf, err = decoder.DecodeAll(it.readBlockBuf, it.readBuf[:0])
+
+	if err != nil {
+		log.Errorf("decodeWALBlock: decompression failed: %v", err)
+		return err
+	}
+
+	rawReader := bytes.NewReader(it.readBuf)
+
+	var N uint32
+	err = binary.Read(rawReader, binary.LittleEndian, &N)
+	if err != nil {
+		log.Errorf("decodeWALBlock: failed to read datapoint count: %v", err)
+		return err
+	}
+
+	it.readDps = toputils.ResizeSlice(it.readDps, int(N+uint32(len(it.readDps))))
```

Comment:
everytime you read a new block, you have to reuse the it.readDps , you don't need to maintain the old datapoints. other wise you will reun out of memory.